File: sdk/java/src/test/java/jp/kengos/shojiku/VerificationTest.java
package jp.kengos.shojiku;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/**
 * Verification, and the no-false-assurance rule that shapes it.
 *
 * <p>A signature that does not verify is a FAILED result — so a caller who checks only success is
 * never told a forgery is fine — and the report rides that failed result, because {@code
 * notChecked} must reach the caller either way.
 */
class VerificationTest extends Fixtures {

  @Test
  void aGoodSignatureVerifiesAgainstTheCertificateThatSignedIt() {
    ShojikuClient client = client().build();

    Result<VerificationReport> result =
        client.verify(signed(client), List.of(key("rsa2048.cert.pem")), null);

    assertTrue(result.success());
    assertNotNull(result.report());
    assertTrue(result.report().valid());
    assertTrue(result.report().signature().passed());
    assertTrue(result.report().coverage().passed());
  }

  @Test
  void thePassingVerdictStillNamesWhatWasNotChecked() {
    // The no-false-assurance rule on the PASSING path, which is the one a binding
    // is most likely to quietly drop.
    ShojikuClient client = client().build();

    VerificationReport report =
        client.verify(signed(client), List.of(key("rsa2048.cert.pem")), null).unwrap();

    assertEquals(List.of("revocation", "timestamp"), report.notChecked());
  }

  @Test
  void theFailingVerdictNamesThemToo() {
    // The other half, and the whole point of carrying notChecked: it must reach
    // the caller either way, so the report rides a FAILED result too.
    ShojikuClient client = client().build();

    Result<VerificationReport> result =
        client.verify(tampered(client), List.of(key("rsa2048.cert.pem")), null);

    assertTrue(result.failed());
    assertNotNull(result.report());
    assertEquals(List.of("revocation", "timestamp"), result.report().notChecked());
  }

  @Test
  void alteredBytesFailTheResultAndSayWhichCheck() {
    // The four checks stay separate: "valid but covers only part of the file" is
    // a different fact from "the signature is wrong".
    ShojikuClient client = client().build();

    Result<VerificationReport> result =
        client.verify(tampered(client), List.of(key("rsa2048.cert.pem")), null);

    assertTrue(result.failed());
    assertFalse(result.report().valid());
    assertFalse(result.report().signature().passed());
    assertTrue(result.report().coverage().passed());
    assertEquals("signature", result.failure().kind());
    assertEquals(Step.VERIFY, result.failure().step());
  }

  @Test
  void theFourChecksHaveNamesACallerCanBranchOn() {
    ShojikuClient client = client().build();

    VerificationReport report =
        client.verify(signed(client), List.of(key("rsa2048.cert.pem")), null).unwrap();

    assertEquals(
        List.of("certificateValidity", "coverage", "signature", "trustChain"),
        report.checks().keySet().stream().sorted().toList());
  }

  @Test
  void anAnchorThatSignedNothingHereFailsTheChain() {
    ShojikuClient client = client().build();

    Result<VerificationReport> result =
        client.verify(signed(client), List.of(key("other-ca.cert.pem")), null);

    assertTrue(result.failed());
    assertNotNull(result.report());
    assertFalse(result.report().trustChain().passed());
  }

  @Test
  void aChainIssuedLeafVerifiesAgainstItsAuthority() {
    ShojikuClient client = client().build();
    LocalPem leaf = new LocalPem(key("leaf.key.pem"), key("leaf.cert.pem"));
    DocumentArtifact document = client.sign(rendered(client), leaf).unwrap();

    Result<VerificationReport> result = client.verify(document, List.of(key("ca.cert.pem")), null);

    assertTrue(result.success());
    assertTrue(result.unwrap().trustChain().passed());
  }

  @Test
  void anExpiredCertificateFailsValidityRatherThanTheSignature() {
    ShojikuClient client = client().build();
    LocalPem expired = new LocalPem(key("leaf.key.pem"), key("leaf-expired.cert.pem"));
    DocumentArtifact document = client.sign(rendered(client), expired).unwrap();

    VerificationReport report = client.verify(document, List.of(key("ca.cert.pem")), null).report();

    assertNotNull(report);
    assertFalse(report.certificateValidity().passed());
    assertTrue(report.signature().passed());
  }

  @Test
  void aDocumentWithNoSignatureInItGivesNoReportAtAll() {
    // A document that cannot be evaluated at all has NO report, which is a
    // different fact from an empty one.
    ShojikuClient client = client().build();

    Result<VerificationReport> result =
        client.verify(
            client.artifact(rendered(client).bytes()), List.of(key("rsa2048.cert.pem")), null);

    assertTrue(result.failed());
    assertNull(result.report());
  }

  @Test
  void severalAnchorFilesAreTakenAtOnceAsTheCliTakesSeveralFlags() {
    ShojikuClient client = client().build();

    assertTrue(
        client
            .verify(
                signed(client), List.of(key("other-ca.cert.pem"), key("rsa2048.cert.pem")), null)
            .success());
  }

  @Test
  void anchorsMayBeBytesForACertificateThatNeverTouchedDisk() throws IOException {
    ShojikuClient client = client().build();

    assertTrue(
        client.verify(signed(client), null, Files.readAllBytes(key("rsa2048.cert.pem"))).success());
  }

  @Test
  void unusableAnchorBytesAreAFailedResult() {
    ShojikuClient client = client().build();

    Result<VerificationReport> result =
        client.verify(signed(client), null, "not a pem at all".getBytes(StandardCharsets.US_ASCII));

    assertTrue(result.failed());
    assertEquals("anchors", result.failure().kind());
  }

  @Test
  void anUnreadableAnchorFileIsAFailedResultNotAnException() {
    ShojikuClient client = client().build();

    Result<VerificationReport> result =
        client.verify(signed(client), List.of(Path.of("/nonexistent/anchor.pem")), null);

    assertTrue(result.failed());
    assertEquals("anchor_unreadable", result.failure().kind());
    assertEquals(Step.VERIFY, result.failure().step());
  }

  @Test
  void anchorsAreRequiredBecauseThereIsNoTrustStoreToFallBackOn() {
    // The engine never consults the machine's trust store, so a default would
    // answer a different question than the caller asked.
    ShojikuClient client = client().build();

    UsageException error =
        assertThrows(UsageException.class, () -> client.verify(signed(client), null, null));

    assertTrue(error.getMessage().contains("needs"));
  }

  @Test
  void passingBothAnchorFormsIsProgrammerMisuse() {
    ShojikuClient client = client().build();

    assertThrows(
        UsageException.class,
        () -> client.verify(signed(client), List.of(Path.of("/a.pem")), new byte[] {1}));
  }

  @Test
  void theArtifactCanVerifyItself() {
    ShojikuClient client = client().build();

    assertTrue(signed(client).verify(List.of(key("rsa2048.cert.pem"))).success());
  }

  @Test
  void theArtifactCanVerifyItselfFromAnchorBytes() throws IOException {
    ShojikuClient client = client().build();

    assertTrue(signed(client).verifyPem(Files.readAllBytes(key("rsa2048.cert.pem"))).success());
  }

  @Test
  void anArchivedDocumentVerifiesAfterReEntry() {
    // The whole point of `artifact(bytes)`: a document signed some time ago, read
    // back from wherever it was stored — and verification is never restricted, so
    // even a strict client can check it.
    byte[] archived = signed(client().build()).bytes();
    ShojikuClient strict = client().strict(true).providers(Map.of()).build();

    assertTrue(strict.artifact(archived).verify(List.of(key("rsa2048.cert.pem"))).success());
  }

  /**
   * A signed document with one byte of the ORIGINAL body flipped.
   *
   * <p>Corrupting the middle of the SIGNED file lands in the appended revision instead, which
   * leaves a container the verifier cannot parse a signature out of at all — no report, which is a
   * different outcome from the one these tests are about.
   */
  private static DocumentArtifact tampered(ShojikuClient client) {
    DocumentArtifact rendered = rendered(client);
    byte[] bytes = client.sign(rendered, signer()).unwrap().bytes();
    bytes[rendered.size() / 2] ^= (byte) 0xFF;
    return client.artifact(bytes);
  }
}

File: sdk/php/tests/VerificationTest.php
<?php

declare(strict_types=1);

namespace Shojiku\Tests;

use PHPUnit\Framework\TestCase;
use Shojiku\Exception\UsageException;
use Shojiku\Step;
use Shojiku\VerificationCheck;
use Shojiku\VerificationReport;

final class VerificationTest extends TestCase
{
    use EngineFixtures;

    public function testAValidSignatureVerifiesAgainstItsAnchor(): void
    {
        $result = $this->signed()->verify(anchors: $this->keyPath('rsa2048.cert.pem'));

        self::assertTrue($result->success());
        $report = $result->report();
        self::assertNotNull($report);
        self::assertTrue($report->valid());
        self::assertTrue($report->signature()->passed());
        self::assertTrue($report->coverage()->passed());
        self::assertTrue($report->certificateValidity()->passed());
        self::assertTrue($report->trustChain()->passed());
    }

    public function testWhatWasNotCheckedSurvivesAPassingVerdict(): void
    {
        $report = $this->signed()->verify(anchors: $this->keyPath('rsa2048.cert.pem'))->report();

        self::assertNotNull($report);
        self::assertContains('revocation', $report->notChecked());
        self::assertContains('timestamp', $report->notChecked());
    }

    public function testATamperedDocumentIsAFailedResultThatStillCarriesItsReport(): void
    {
        // The byte is flipped INSIDE the original body, indexed from the
        // pre-signature length: signing APPENDS a revision, so a flip at the
        // file's midpoint would land in the appended part and leave a
        // container the verifier cannot parse a signature out of at all —
        // which is a different outcome from the one this pins.
        $original = $this->rendered()->bytes();
        $tampered = $this->signed()->bytes();
        $at = intdiv(strlen($original), 2);
        // The mask is for the ANALYSER, not the arithmetic: xor-ing two values
        // that are each at most 0xff cannot exceed 0xff, but phpstan does not
        // narrow an integer range across `^`, so on PHP 8.5 it reads the result
        // as a plain int and `chr` now declares `int<0, 255>`.
        $tampered[$at] = chr((ord($tampered[$at]) ^ 0x20) & 0xFF);

        $result = $this->client()->artifact($tampered)
            ->verify(anchors: $this->keyPath('rsa2048.cert.pem'));

        // A signature that does not verify is a FAILED result — a caller who
        // checks only success() is not told a forgery is fine.
        self::assertTrue($result->failed());
        $failure = $result->failure();
        self::assertNotNull($failure);
        self::assertSame(Step::Verify, $failure->step());
        self::assertSame('signature', $failure->kind());

        // …and the report rides that failed result, because `notChecked` has
        // to reach the caller either way.
        $report = $result->report();
        self::assertNotNull($report);
        self::assertFalse($report->valid());
        self::assertFalse($report->signature()->passed());
        self::assertNotNull($report->signature()->reason());
        self::assertTrue($report->coverage()->passed());
        self::assertContains('revocation', $report->notChecked());
    }

    public function testADocumentThatCannotBeEvaluatedHasNoReportAtAll(): void
    {
        // Different from an empty report: nothing was checked, so there is
        // nothing to say about what was not checked.
        $result = $this->rendered()->verify(anchors: $this->keyPath('rsa2048.cert.pem'));

        self::assertTrue($result->failed());
        self::assertNull($result->report());
        self::assertSame('verify', $result->failure()?->kind());
    }

    public function testAnchorsCrossAsPathsOrAsBytes(): void
    {
        $pem = (string) file_get_contents($this->keyPath('rsa2048.cert.pem'));

        self::assertTrue($this->signed()->verify(anchorsPem: $pem)->success());
        self::assertTrue($this->signed()->verify(anchors: [$this->keyPath('rsa2048.cert.pem')])->success());
    }

    public function testAnUntrustedAnchorFailsTheChainRatherThanTheSignature(): void
    {
        $result = $this->signed()->verify(anchors: $this->keyPath('other-ca.cert.pem'));

        self::assertTrue($result->failed());
        $report = $result->report();
        self::assertNotNull($report);
        self::assertTrue($report->signature()->passed());
        self::assertFalse($report->trustChain()->passed());
        // "The signature is valid but nobody we trust vouches for it" is a
        // different fact from "the signature is wrong", and a caller that
        // cannot tell them apart cannot explain the answer to anyone.
        self::assertNotSame('', (string) $report->trustChain());
    }

    public function testBothAnchorFormsAtOnceIsProgrammerMisuse(): void
    {
        $this->expectException(UsageException::class);
        $this->expectExceptionMessageMatches('/either `anchors:` \(paths\) or `anchorsPem:` \(bytes\), not both/');
        $this->signed()->verify(anchors: '/a.pem', anchorsPem: 'bytes');
    }

    public function testNoAnchorAtAllIsProgrammerMisuse(): void
    {
        // There is deliberately no fallback to the machine's trust store: the
        // engine never consults one, so a default would answer a different
        // question than the caller asked.
        $this->expectException(UsageException::class);
        $this->expectExceptionMessageMatches('/verify needs/');
        $this->signed()->verify();
    }

    public function testAnEmptyAnchorListIsProgrammerMisuseToo(): void
    {
        // The same statement as none at all — and if it reached the engine it
        // would come back as a refused INVOCATION, which a caller would read
        // as a transport failure rather than as the misuse it is.
        $this->expectException(UsageException::class);
        $this->expectExceptionMessageMatches('/verify needs/');
        $this->signed()->verify(anchors: []);
    }

    public function testAnUnreadableAnchorIsAFailedResult(): void
    {
        $result = $this->signed()->verify(anchors: $this->keyPath('no-such-anchor.pem'));

        self::assertTrue($result->failed());
        self::assertSame('io', $result->failure()?->kind());
    }

    public function testACheckWithoutAReasonPrintsItsStatusAlone(): void
    {
        $passed = VerificationCheck::parse(['status' => 'passed']);
        $failed = VerificationCheck::parse(['status' => 'failed', 'reason' => 'the digest differs']);

        self::assertSame('passed', (string) $passed);
        self::assertNull($passed->reason());
        self::assertSame('failed: the digest differs', (string) $failed);
        self::assertFalse($failed->passed());
    }

    public function testAMalformedCheckIsEmptyRatherThanGuessedAt(): void
    {
        $check = VerificationCheck::parse('not an object');

        self::assertNull($check->status());
        self::assertFalse($check->passed());
        self::assertSame('', (string) $check);
    }

    public function testAReportExposesItsChecksAsAMapAsWellAsSeparately(): void
    {
        $report = VerificationReport::parse([
            'valid' => true,
            'signature' => ['status' => 'passed'],
            'coverage' => ['status' => 'passed'],
            'certificateValidity' => ['status' => 'passed'],
            'trustChain' => ['status' => 'passed'],
            'notChecked' => ['revocation', 42],
        ]);

        self::assertSame(
            ['signature', 'coverage', 'certificateValidity', 'trustChain'],
            array_keys($report->checks()),
        );
        // A non-string entry is dropped rather than cast: the list is the
        // engine's vocabulary, not free text.
        self::assertSame(['revocation'], $report->notChecked());
    }

    public function testAReportWithoutAValidFlagIsNotValid(): void
    {
        $report = VerificationReport::parse([]);

        self::assertFalse($report->valid());
        self::assertSame([], $report->notChecked());
    }

    public function testALockedDownClientCanStillVerifyAnythingItIsGiven(): void
    {
        // Verification is NEVER restricted: verifying bytes of unknown
        // provenance is the point of verify, and a locked-down deployment is
        // precisely the one that must check an archived document it did not
        // produce.
        $strict = $this->client(['strict' => true]);
        $archived = $strict->artifact($this->signed()->bytes());

        self::assertTrue($archived->verify(anchors: $this->keyPath('rsa2048.cert.pem'))->success());
    }
}
